component-api/src/main/java/org/talend/sdk/component/api/record/OrderedMap.java

```diff
@@ -0,0 +1,310 @@
+/**
+ * Copyright (C) 2006-2022 Talend Inc. - www.talend.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.talend.sdk.component.api.record;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/**
+ * On TCK Record, we have to control order of elements with keeping efficient access.
+ * LinkedHashMap has efficient access, but order of element is only the order of insertions.
+ * List (ArrayList, LinkedList) allow to fine control order but have inefficent access.
+ * This class aims to control element order with keeping efficient access.
+ * 
+ * @param <T> : type of element.
+ */
+public class OrderedMap<T> {
```

Comment:
Does it have to be public? It feels to me that it's an implementation detail of the default `*Impl` implementation of `Schema`?

Also, since it does not implement `Map`, I think it feels a bit weird to have it named `OrderedMap`. If this becomes package-private this is not an issue, but if it has to stay public, it could become a problem, WDYT?

Reply:
For "public", as it is used in api module itself (Schema), and also in 2 others (spi and impl), it can't be other than public. 

About name, i choose "Map" because it establish a relation between a key (the name) and an object (T), but indeed, we may find better, make suggestions.

Reply:
I think that this class includes some ambiguous namings (entry,entries, schemaEntry,etc.)!
Many member names seem to come from Schema interface.
I suggest to be more generic on member naming for this class. (Node is fine for me in contrast)

Reply:
I agree w/ Nicolas, not sure it has to be named `XxxxMap`...  
Naming is hard sometimes ;-)